src/server/ApiDispatchers.cpp

```diff
@@ -1252,38 +1252,35 @@ static DWORD TraceGetThreadId(CONSOLE_API_MSG* const m)
     ULONG cbBufferSize;
     RETURN_IF_FAILED(m->GetInputBuffer(&pvBuffer, &cbBufferSize));
 
-    PVOID pvInputTarget;
-    const auto cbInputTarget = a->TargetLength;
-    PVOID pvInputExeName;
-    const auto cbInputExeName = a->ExeLength;
-    PVOID pvInputSource;
-    const auto cbInputSource = a->SourceLength;
-    // clang-format off
-    RETURN_HR_IF(E_INVALIDARG, !IsValidStringBuffer(a->Unicode,
-                                                    pvBuffer,
-                                                    cbBufferSize,
-                                                    3,
-                                                    cbInputExeName,
-                                                    &pvInputExeName,
-                                                    cbInputSource,
-                                                    &pvInputSource,
-                                                    cbInputTarget,
-                                                    &pvInputTarget));
-    // clang-format on
+    // There are 3 strings stored back-to-back within the message payload.
+    // First we verify that their size and alignment are alright and then we extract them.
+    const ULONG cbInputExeName = a->ExeLength;
+    const ULONG cbInputSource = a->SourceLength;
+    const ULONG cbInputTarget = a->TargetLength;
+
+    const auto alignment = a->Unicode ? alignof(wchar_t) : alignof(char);
+    // ExeLength, SourceLength and TargetLength are USHORT and summing them up will not overflow a ULONG.
+    const auto badLength = cbInputTarget + cbInputExeName + cbInputSource > cbBufferSize;
+    const auto badAlignment = ((cbInputExeName | cbInputSource | cbInputTarget) & (alignment - 1)) != 0;
+    RETURN_HR_IF(E_INVALIDARG, badLength || badAlignment);

```

Comment:
These 3 lines in particular are a bit "meh". If you have a better idea, let me know!